src/constructs/ServerSideWebsite.ts

```diff
@@ -0,0 +1,430 @@
+        if (bucketName === undefined) {
+            // No bucket found => nothing to delete!
+            return;
+        }
+
+        log(
+            `Emptying S3 bucket '${bucketName}' for the '${this.id}' website, else CloudFormation will fail (it cannot delete a non-empty bucket)`
+        );
+        await emptyBucket(this.provider, bucketName);
+    }
+
+    async getUrl(): Promise<string | undefined> {
+        const domain = await this.getDomain();
+        if (domain === undefined) {
+            return undefined;
+        }
+
+        return `https://${domain}`;
+    }
+
+    async getBucketName(): Promise<string | undefined> {
+        return this.provider.getStackOutput(this.bucketNameOutput);
+    }
+
+    async getDomain(): Promise<string | undefined> {
+        return this.provider.getStackOutput(this.domainOutput);
+    }
+
+    async getCName(): Promise<string | undefined> {
+        return this.provider.getStackOutput(this.cnameOutput);
+    }
+
+    async getDistributionId(): Promise<string | undefined> {
+        return this.provider.getStackOutput(this.distributionIdOutput);
+    }
+
+    getMainCustomDomain(): string | undefined {
+        if (this.configuration.domain === undefined) {
+            return undefined;
+        }
+
+        // In case of multiple domains, we take the first one
+        return typeof this.configuration.domain === "string" ? this.configuration.domain : this.configuration.domain[0];
+    }
+
+    private headersToForward(): OriginRequestHeaderBehavior {
+        let additionalHeadersToForward = this.configuration.forwardedHeaders ?? [];
+        if (additionalHeadersToForward.includes("Host")) {
+            throw new ServerlessError(
+                `Invalid value in 'constructs.${this.id}.forwardedHeaders': the 'Host' header cannot be forwarded (this is an API Gateway limitation). Use the 'X-Forwarded-Host' header in your code instead (it contains the value of the original 'Host' header).`,
+                "LIFT_INVALID_CONSTRUCT_CONFIGURATION"
+            );
+        }
+        // `Authorization` cannot be forwarded via this setting (we automatically forward it anyway so we remove it from the list)
+        additionalHeadersToForward = additionalHeadersToForward.filter((header: string) => header !== "Authorization");
+
+        /**
+         * We forward everything except:
+         * - `Host` because it messes up API Gateway (that uses the Host to identify which API Gateway to invoke)
+         * - `Authorization` because it must be configured on the cache policy
+         *   (see https://aws.amazon.com/premiumsupport/knowledge-center/cloudfront-authorization-header/?nc1=h_ls)
+         */
+        return OriginRequestHeaderBehavior.allowList(
+            "Accept",
+            "Accept-Language",
+            "Origin",
+            "Referer",
+            // This header is set by our CloudFront Function
+            "X-Forwarded-Host",
```

Comment:
> Do you see any other header we should probably include?

I'm not sure what's the best approach here.

I think getting a list right is hard and the list of standard headers is very long. And some non-standard headers are more widespread than standard ones. 

For example, `x-api-key` or `x-forwarded-*` headers are non standard, but their usage is probably higher than `Range` or `From` standard headers.

My point here is that any list we may provide will be biaised and the more we add, the more people will either want us to add more to the list, or ask us to remove some.

So i see two solutions 
1. We stay on the same track and maybe add some more headers to the mix like, but not limited to, CORS related, `X-Requested-With`, all `Accept-*`, `User-Agent`, all `Content-*`
2. We allow every headers to be forwarded (maybe even include Cloudfront headers). This is less restrictive but is generally what is done by default by web servers like Apache or Nginx. We don't get in the way and this feels familiar for most users. We may still support a `forwardedHeaders` list but the user would be expected to provide each and every header they need : they choose to be picky and must know what they are doing. 

Let me know what you think

Reply:
> We allow every headers to be forwarded

That would be the ideal scenario, but we cannot forward the `Host` header because it breaks API Gateway. (indeed, API Gateway uses the Host header to identify which API to call)

Ideally, we would "forward all except Host" but I don't think it's possible to exclude headers.

There's also [this extreme solution](https://stackoverflow.com/a/53343079/245552), but it requires using a custom domain. And it's a bit weird.

Reply:
(FYI I've asked Eric Johnson from AWS, he's an API Gateway expert, maybe he has some ideas)

Reply:
And what about unsetting the `host` header in the Cloudfront Function ? Would that work ?

That way we would allow all headers and remove just the `host` header.

Reply:
Just tested this and `Host` is a read-only header so we're unable to delete it in a Cloudfront function :frowning: 

https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/edge-functions-restrictions.html#function-restrictions-read-only-headers-viewer-request

Reply:
And on top of that, `OriginRequestHeaderBehavior.allowList` is limited to 10 headers. We're already setting 5 of them which leaves 5 headers for the end user to configure. That sucks but I don't see any other way to do this, that's an AWS limitation.

Let's wait for the feedback from Eric but if he has no proper solution, I think the current implementation is the way to go. And maybe add an Error when the user specify more than 5 custom headers (the CDK doesn't validate it, you'd get a Cloudformation error at deploy).

Reply:
No solution from the API Gateway team unfortunately :( 

> And on top of that, OriginRequestHeaderBehavior.allowList is limited to 10 headers.

This is aweful, where did you read that? I can't find any mention of it in the docs:

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cloudfront-originrequestpolicy-headersconfig.html#cfn-cloudfront-originrequestpolicy-headersconfig-headers

I keep coming back to this thinking this isn't ideal at all (to provide a generic component). Debugging why some headers are not received by a backend app will waste hours for developers (they'll never think about the serverless.yml config).

The `10` limit makes it much more complex: ideally I would want to whitelist the most common headers by default (even go up to 10 headers). But then if we make it configurable, that means user cannot add new headers, they would have to redefine the list completely to fit their needs. Which isn't great UX :/ I'm not sure what to do here TBH.

Reply:
> This is aweful, where did you read that?

I tested some things out in the console and it's stated there. And that limit is also present when using Cloudformation. Didn't found useful docs either.

I think even with the 10 headers limit, it still has a great value. From my experience, most apps don't need that much headers. 

I agree on the UX part and don't see an easy way to avoid this pitfall. Maybe one way could be to force users to write the list of headers and make it a required field. That way, it makes it explicit that headers need to be configured and we may provide a default list in the docs. Not ideal, I agree...

I check on serverlessland and found that they have a similar pattern implemented with CDK and don't worry about headers there apparently. I haven't looked at it in details but maybe it contains something valuable (or maybe they just don't care about headers...). Here it is : https://serverlessland.com/patterns/cloudfront-s3-lambda-cdk

Let me know if I can be of any help.

Reply:
OK thanks a lot @t-richard for all your advice.

You're right, the construct can still be very helpful to a lot of people, let's not stop on that.

What I did is:

- add `X-Requested-With` to the default list
- make the YAML option an "override" instead of an "extend" -> that gives users control over 10 headers, instead of just 4. Later if needed we can add a `additionalHeadersToForward` option.

Last question on my mind: are there any other obvious header to forward? CloudFront has separate settings for cookies for examples, I need to test if they actually work correctly.

Except that, I think this problem is finally solved.